### ISSUE TITLE
Context leak in Tomcat

### ISSUE THREAD

**Issue by iNikem** [open-telemetry/opentelemetry-java-instrumentation#787]
`TomcatServlet3TestDispatchAsync.test success with 4 requests` test fails constantly in Github Action https://github.com/open-telemetry/opentelemetry-java-instrumentation/actions?query=workflow%3A%22Nightly+build%22 . The reason for a failure can be seen from this log:
```

13:55:36.123 [Test worker] DEBUG http-client - --> GET http://localhost:36897/tomcat-context/dispatch/success |  
-- | --
  | 13:55:36.124 [http-nio-36897-exec-3] DEBUG io.opentelemetry.auto.tooling.Instrumenter$Default - Applying instrumentation: java_concurrent -- io.opentelemetry.auto.instrumentation.javaconcurrent.RunnableInstrumentation on jdk.internal.loader.ClassLoaders$AppClassLoader@5c29bfd |  
  | 13:55:36.129 [http-nio-36897-exec-3] DEBUG io.opentelemetry.auto.tooling.context.FieldBackedProvider - Found context-store access in io.opentelemetry.auto.instrumentation.javaconcurrent.RunnableInstrumentation |  
  | 13:55:36.131 [http-nio-36897-exec-3] DEBUG io.opentelemetry.auto.tooling.context.FieldBackedProvider - Rewriting context-store map fetch for instrumenter io.opentelemetry.auto.instrumentation.javaconcurrent.RunnableInstrumentation: java.lang.Runnable -> io.opentelemetry.auto.bootstrap.instrumentation.java.concurrent.State |  
  | 13:55:36.132 [http-nio-36897-exec-3] DEBUG io.opentelemetry.auto.tooling.AgentInstaller$TransformLoggingListener - Transformed org.apache.tomcat.util.net.NioBlockingSelector$BlockPoller$RunnableRemove -- jdk.internal.loader.ClassLoaders$AppClassLoader@5c29bfd |  
  | 13:55:36.143 [http-nio-36897-exec-3] DEBUG io.opentelemetry.auto.test.InMemoryExporter - >>> SPAN START: TEST_SPAN id=70ec8a85d67d75a7 traceid=d559ee1f62c109196811925b12ae92c0 parent=f42a8abecfe071db, library=InstrumentationLibraryInfo{name=io.opentelemetry.auto, version=null} |  
  | <<< SERVER SPAN END: ApplicationFilterChain.doFilter library=InstrumentationLibraryInfo{name=io.opentelemetry.auto.servlet-3.0, version=null}, id=f42a8abecfe071db traceid=d559ee1f62c109196811925b12ae92c0 parent=0000000000000000, attributes=Attribute http.status_code=200Attribute net.peer.port=53844Attribute servlet.path=/dispatch/successAttribute servlet.context=/tomcat-contextAttribute http.url=http://localhost:36897/tomcat-context/dispatch/successAttribute net.peer.ip=127.0.0.1Attribute http.method=GET |  
  |   |  
  | 13:55:36.149 [http-nio-36897-exec-3] DEBUG io.opentelemetry.auto.test.InMemoryExporter - >>> SPAN START: ApplicationFilterChain.doFilter id=64de582154ff3af3 traceid=ebd90169d9db81088c3e3ac6c2fb2f35 parent=0000000000000000, library=InstrumentationLibraryInfo{name=io.opentelemetry.auto.servlet-3.0, version=null} |  
  | 13:55:36.150 [http-nio-36897-exec-3] DEBUG io.opentelemetry.auto.test.InMemoryExporter - <<< SPAN END: TEST_SPAN id=70ec8a85d67d75a7 traceid=d559ee1f62c109196811925b12ae92c0 parent=f42a8abecfe071db, attributes= |  
  | 13:55:36.151 [http-nio-36897-exec-6] DEBUG io.opentelemetry.auto.test.InMemoryExporter - >>> SPAN START: controller id=346e317ce978b7e8 traceid=ebd90169d9db81088c3e3ac6c2fb2f35 parent=64de582154ff3af3, library=InstrumentationLibraryInfo{name=io.opentelemetry.auto, version=null} |  
  | 13:55:36.151 [http-nio-36897-exec-6] DEBUG io.opentelemetry.auto.test.InMemoryExporter - <<< SPAN END: controller id=346e317ce978b7e8 traceid=ebd90169d9db81088c3e3ac6c2fb2f35 parent=64de582154ff3af3, attributes= |  
  | 13:55:36.151 [http-nio-36897-exec-3] DEBUG io.opentelemetry.auto.test.InMemoryExporter - <<< SPAN END: ApplicationFilterChain.doFilter id=64de582154ff3af3 traceid=ebd90169d9db81088c3e3ac6c2fb2f35 parent=0000000000000000, attributes=Attribute http.status_code=200Attribute net.peer.port=53844Attribute servlet.path=/dispatch/successAttribute servlet.context=/tomcat-contextAttribute http.url=http://localhost:36897/tomcat-context/dispatch/successAttribute net.peer.ip=127.0.0.1Attribute http.method=GET |  
  | 13:55:36.151 [Test worker] DEBUG http-client - <-- 200 http://localhost:36897/tomcat-context/dispatch/success (27ms, 7-byte body)


```

As you see, `TEST_SPAN`, created by `HttpServerTestAdvice.ServerEntryAdvice`, has a parent span, which is unexpected. It has to be the very span created for a given request during test. It can also be seen from the log above, that this unexpected span was created by the servlet instrumentation, which we are now testing. I assume this is a span left from the previous test run `TomcatServlet3TestDispatchAsync.test success with 1 requests`.

After digging, I have discovered that Tomcat, when processing async request, may do the following. First, we see test request processing:
```
	at io.opentelemetry.auto.bootstrap.instrumentation.java.concurrent.ExecutorInstrumentationUtils.setupState(ExecutorInstrumentationUtils.java:77)
	at java.base/java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1319)
	at org.apache.tomcat.util.threads.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:167)
	at org.apache.tomcat.util.threads.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:145)
	at org.apache.tomcat.util.net.SocketWrapperBase.execute(SocketWrapperBase.java:150)
	at org.apache.coyote.AbstractProcessor.execute(AbstractProcessor.java:182)
	at org.apache.coyote.AsyncStateMachine.asyncRun(AsyncStateMachine.java:467)
	at org.apache.coyote.AbstractProcessor.action(AbstractProcessor.java:540)
	at org.apache.coyote.Request.action(Request.java:432)
	at org.apache.catalina.core.AsyncContextImpl.start(AsyncContextImpl.java:245)
	at javax.servlet.AsyncContext$start.call(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:47)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:115)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:127)
	at TestServlet3$DispatchAsync.service(TestServlet3.groovy:160)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:741)
```

`TestServlet3$DispatchAsync.service` creates a span which is later preserved with the `Runnable` when that is passed to the `ThreadPoolExecutor`. During the execution of that runnable, the following stack trace was taken.

```
at io.opentelemetry.auto.bootstrap.instrumentation.java.concurrent.ExecutorInstrumentationUtils.setupState(ExecutorInstrumentationUtils.java:77)
	at java.base/java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1319)
	at org.apache.tomcat.util.threads.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:167)
	at org.apache.tomcat.util.threads.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:145)
	at org.apache.tomcat.util.net.AbstractEndpoint.processSocket(AbstractEndpoint.java:1085)
	at org.apache.tomcat.util.net.SocketWrapperBase.processSocket(SocketWrapperBase.java:812)
	at org.apache.coyote.AbstractProcessor.processSocketEvent(AbstractProcessor.java:782)
	at org.apache.coyote.AbstractProcessor.action(AbstractProcessor.java:503)
	at org.apache.coyote.Request.action(Request.java:430)
	at org.apache.catalina.core.AsyncContextImpl.dispatch(AsyncContextImpl.java:212)
	at org.apache.catalina.core.AsyncContextImpl.dispatch(AsyncContextImpl.java:175)
	at javax.servlet.AsyncContext$dispatch$0.call(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:47)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:115)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:127)
	at TestServlet3$DispatchAsync$_service_closure1.doCall(TestServlet3.groovy:161)
	at TestServlet3$DispatchAsync$_service_closure1.doCall(TestServlet3.groovy)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:101)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:323)
	at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:263)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1041)
	at groovy.lang.Closure.call(Closure.java:405)
	at groovy.lang.Closure.call(Closure.java:399)
	at groovy.lang.Closure.run(Closure.java:486)
	at org.apache.catalina.core.AsyncContextImpl$RunnableWrapper.run(AsyncContextImpl.java:534)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

It is important to note, that this stack trace has an active context which was associated with the Runnable passed from the previous stack trace. Meaning that this context has a valid Span. At the end of this stack the async request was successfully processed  (in `org.apache.coyote.AbstractProcessor.action(AbstractProcessor.java:503)`) and Tomcat moves forward. It now submits next Runnable to process new data from the socket. But we will pass current context to the this Runnable as well! With valid span in it. Thus, when that Runnable will activates on another thread and will process new request, this new request will have active context with valid span in it. Thus we have leaked span.

**Comment by iNikem** [comment]
I am still unable to reproduce that locally. But it happens with 100% probability during GitHub Action.

**Comment by trask** [comment]
Nice find!

What do you think of adding instrumentation on `AsyncContext.start(Runnable)` that propagates the context directly to its Runnable (somehow), and then creates empty Scope (Context), so that Context won't propagate any further down and pollute the servlet container's thread pool?

And also create empty Scope (Context) inside of `AsyncContext.dispatch()` to not pollute servlet container thread pool also (we already have different mode of propagation in that case).

**Comment by trask** [comment]
I guess the real question (that you are asking) is why the Context is getting stuck in the thread pool threads in the first place, since the Runnables which run in those thread pool threads should start/close their scopes and not leave anything behind for future workers.

Makes sense to get to the bottom of that instead of just hiding the problem via my ideas above.

Carry on 😄👍

**Comment by iNikem** [comment]
Here is an example:
```
public class Main {
  private static final Executor pool = Executors.newFixedThreadPool(5);

  public static void main(String[] args) {
    Tracer tracer = OpenTelemetry.getTracer("run.leak");
    Span span = tracer.spanBuilder("first").startSpan();
    try (Scope ignored = tracer.withSpan(span)) {
      Runnable seed = new Work();
      pool.execute(seed);
    } catch (Exception e) {
      e.printStackTrace();
    }
  }

  private static class Work implements Runnable {

    @Override
    public void run() {
      System.out.println("Do work");
      try {
        TimeUnit.MILLISECONDS.sleep(100);
      } catch (InterruptedException e) {
        System.out.println("Interrupted");
        Thread.currentThread().interrupt();
      }
      Runnable nextJob = new Work();
      pool.execute(nextJob);
    }
  }
}
```
The key element here is that every `run` method creates and executes next unit work. As `pool.execute` always happens _inside_ the current scope, either manually created or by auto-instrumentation, all Runnables carry the same span context.

Seems like "works as expected". But if this is similar to how Tomcat works (I believe so at the moment), then some of those workers actually start processing new incoming http request. And this leads to the problem above.

**Comment by trask** [comment]
@iNikem can we close this now that we've decided #814 is a reasonable (permanent) fix for this?

**Comment by iNikem** [comment]
@trask what do you think about https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/814#issuecomment-669713447 ?

**Comment by iNikem** [comment]
Fixed by #814 

After some discussions we decided to use framework-specific exclusions as a way forward.
